### PR TITLE
HDBSCAN-BC

### LINKED ISSUE #87: Implement semi-supervised clustering with HDBSCAN
HDBSCAN's [semi-supervised extension](https://link.springer.com/article/10.1007/s10618-019-00651-1) would be a great addition to the library. For example, given a partially-labeled data (e.g. 10-20% of points are labeled by the user), we would like to obtain clustering results that best fits the points of the same class into the same clusters as well as discover novel clusters based on density.

Currently, there are two implementations of semi-supervised clustering using HDBSCAN, in [Java](https://github.com/jadsoncastro/UnifiedView) and [Python](https://github.com/TutteInstitute/fast_hdbscan). The Python version is quite recent, but its minimum spanning tree looks to be approximate. So I was thinking of implementing the framework for `HDbscan`, based on either of the following options. 

### Option A.
[`Predict`](https://github.com/petabi/petal-clustering/blob/7e077bd2c9d1e5bc539468f185b2f68b51f3ec44/src/lib.rs#L18) trait can be changed to the following and then implemented by HDBSCAN:
```
/// An interface to apply a trained model.
pub trait Predict<I, O>
where
    I: ?Sized,
{
    fn predict(&mut self, input: &I, partial_labels: O) -> O;
}
```
But this change may not follow the same structure of the Python [implementation](https://github.com/TutteInstitute/fast_hdbscan/blob/519fb3f3feb2f5488c830ea177a4ed5887cd5252/fast_hdbscan/hdbscan.py#L318), where they added `partial_labels` as optional parameter to the `fit` function itself.

### Option B.
Change `Fit` trait by adding `fit_predict` function as in scikit-learn's [DBSCAN](https://github.com/scikit-learn/scikit-learn/blob/2c0cdd479a2d229e88afbb97cce91eef3d1b07c9/sklearn/cluster/_dbscan.py#L448):
```
/// An interface to train a model.
pub trait Fit<I, O>
where
    I: ?Sized,
{
    fn fit(&mut self, input: &I) -> O;
    fn fit_predict(&mut self, input: &I, partial_labels: &O) -> O;
}
```
This will force unnecessary changes in `Optics` or `Dbscan` structs to implement the `fit_predict` function.

### Option C:
Add a new trait called `FitPredict` as in scikit-learn's [HDBSCAN](https://github.com/scikit-learn/scikit-learn/blob/2c0cdd479a2d229e88afbb97cce91eef3d1b07c9/sklearn/cluster/_hdbscan/hdbscan.py#L882):
```
/// An interface to train and predict: 
pub trait FitPredict<I, O>
where
    I: ?Sized,
{
    fn fit_predict(&mut self, input: &I, partial_labels: &O) -> O;
}
```
This will not introduce unnecessary forced changes in other structs and does seem semantically correct to me: we are fitting the unknown data and predicting their clusters based on partial labels.

Please let me know if you would be interested in including semi-supervised clustering for HDBSCAN, and which of the aforementioned options would be the best way to go.

### PR DESCRIPTION
**Pull request by azizkayumov**
Closes #87

### CODE REVIEW

**Comment by azizkayumov** [comment]
@gemini-code-assist please review again and be more specific about the issues you mentioned above.

**Comment by azizkayumov** [comment]
@gemini-code-assist , here is a summary of changes based on your feedback:
FILE: src/dbscan.rs: 
1. Added a docs comment noting that the optional parameter `_params` is not used, but kept for consistency.
FILE: src/hdbscan.rs:
1. Added a docs section to note that the optional parameter is used as partial labels, also noted the algorithms being used if `partial_labels` are provided by the user. 
2 and 3 are the same: refactored `mst` and `union_find` related code into their own modules so that `hdbscan` is easier to manage.
Can you review again?

**Review by msk — APPROVED** [review]
